### PR TITLE
Generate external boot image to launch container image as a VM

### PR DESCRIPTION
**Pull request by zedi-pramodh**
In kvm based eve we launch container as a VM.
To support similar functionality in kubevirt eve we use a feature in kubevirt Look at https://kubevirt.io/user-guide/virtual_machines/boot_from_external_source/ We provide a tiny scratch container image which just consists of kernel and initrd Then the user container image will be converted to a PV/PVC and passed into virtualmachineinstance spec as a boot disk.
EVE mount_disk.sh will mount the PVC as a rootfs in the VM

The code to create virtualmachineinstance spec will come in later commits (domainmgr and kubevirt.go)

This particular comit does the following.
1) Extracts kernel and initrd binaries from eve-kernel and eve-xen-tools packages and creates a scratch container.
2) Converts the container to external-boot-image.tar using docker save
3) Ships external-boot-image.tar in eve-kube container. 
4) When device boots external-boot-image.tar is imported into eve user containerd repo.

Basically every  kubevirt eve build gets its own image, this works nicely for upgrade scenario too. This adds additional 15MB to eve image size which is minimal.

Testing done
============
Compiled and verified nothing is broken in master branch.
Actual launch of container as a VM was tested in kubevirt eve poc branch
On eve device

ctr -a /run/containerd-user/containerd.sock image ls | grep external

docker.io/lfedge/eve-external-boot-image:latest                                                                                                  application/vnd.docker.distribution.manifest.v2+json      sha256:653fb9540cef1c514a428377c4ee28d7ca6d451dabc2cc467e9a5bf0f8576568 13.9 MiB  linux/amd64                                                                   io.cri-containerd.image=managed

### CODE REVIEW

**Review by eriknordmark — COMMENTED** [review]
> Basically every kubevirt eve build gets its own image, this works nicely for upgrade scenario too. This adds additional 15MB to eve image size which is minimal.

Are we adding this to the kvm/xen/kubevirt images or only to the kubevirt images?
I don't know if we want to get 15 MB closer to the 300 MB limit we have to be able to upgrade kvm/xen images.

File: Makefile [hunk 1]
```diff
@@ -757,6 +757,13 @@ pkgs: RESCAN_DEPS=
 pkgs: build-tools $(PKGS)
 	@echo Done building packages
 
+pkg/external-boot-image: eve-external-boot-image
```
Comment (rene) [inline Makefile:760]: @zedi-pramodh , is there any particular reason to create this intermediary external-boot-image.tar tarball image? Because from this PR I don't understand why this file is needed.... IMO you could let everything inside the containers, like this:

1. You've created the pkg/external-boot-image container to extract from eve-kernel the kernel and initrd images
2. You could create the tarball directly from the Dockerfile (of pkg/external-boot-image) and put it inside this container
3. In the Dockerfile of pkg/kube you can just import the generated tarball directly from the pkg/external-boot-image container

In this way you can avoid creating two new targets in the Makefile that will need to be maintained and letting the tarball generated outside the build system, even though you still need the tarball externally, you can export containers from linuxkit cache using the `make export-cache IMAGE=.... OUTFILE=....` target.... but as mentioned, I don't know the whole context and if there is a particular reason to keep this intermediary file....

Comment (rene) [reply Makefile:760]: Thinking about what @eriknordmark commented in this PR, another option would be generate the tarball during runtime, since you will always have both files that you need: kernel + initrd... this approach would:

1. Avoid the creation of the new container
2. Avoid the increasing of the rootfs
3. Avoid problems with different architectures/kernel version, since you will always pick the working (running) kernel + initrd

Drawbacks: You will need to create the tarball (at least once) during runtime, which I think is not a problem at all...

File: pkg/external-boot-image/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# syntax=docker/dockerfile-upstream:1.5.0-rc2-labs
+
+FROM lfedge/eve-xen-tools:e36677ad441c2d5cd88cdd9831959eaa4715ab82 as initrd-build
+FROM docker.io/lfedge/eve-kernel:eve-kernel-amd64-v6.1.38-generic-d2da815271a2-gcc as  kernel-build
```
Comment (rene) [inline pkg/external-boot-image/Dockerfile:4]: You are fixing the kernel architecture + version here, which breaks the compatibility with arm64... so far I can build `make ZARCH=arm64 HV=kubevirt eve-kube/eve/etc` so you need to check the architecture here and pick the right kernel, or at least forbid the build of kubevirt variant for arm64....

**Review by OhmSpectator — COMMENTED** [review]
Could we avoid "Fix Yetus" commits? It's much cleaner to do `git commit --amend --no-edit` or a proper `git rebase -i` if necessary...

**Comment by zedi-pramodh** [comment]
> > Basically every kubevirt eve build gets its own image, this works nicely for upgrade scenario too. This adds additional 15MB to eve image size which is minimal.
> 
> Are we adding this to the kvm/xen/kubevirt images or only to the kubevirt images? I don't know if we want to get 15 MB closer to the 300 MB limit we have to be able to upgrade kvm/xen images.

This is only for Kubevirt images. This package does not get compiled in kvm/xen.  It gets triggered only with make HV=kubevirt compile flag.

File: pkg/external-boot-image/Dockerfile (hunk 2 above)
Comment (zedi-pramodh) [reply pkg/external-boot-image/Dockerfile:4]: For now we have no plans to support in arm64. But I agree we need to have a better of of handling that ? I am thinking of actually passing the KERNEL_TAG to the docker file as env variable in that way architecture will always be correct. I will look into that.

File: Makefile (hunk 1 above)
Comment (zedi-pramodh) [reply Makefile:760]: @rene I am using new container to generate a tarball which gets shipped in kubevirt images only.

The tarball is a docker image tarball (not just collection of kernel and initrd binaries), since we do not have docker in the runtime of the eve, we have to do at build time.

But I liked your comment on eliminating extra Makefile target to achieve the same result.  Would like to learn more on that.

Comment (rene) [reply Makefile:760]: @zedi-pramodh , as I said, you can use the cache-export from our Makefile instead of call docker directly and export the external-boot-image directly in the pkg/kube target, something like this:

```diff
diff --git a/Makefile b/Makefile
index 9e204372a..02a9f849c 100644
--- a/Makefile
+++ b/Makefile
@@ -362,7 +362,7 @@ PKGS=pkg/alpine $(PKGS_$(ZARCH))
 
 # these are the packages that, when built, also need to be loaded into docker
 # if you need a pkg to be loaded into docker, in addition to the lkt cache, add it here
-PKGS_DOCKER_LOAD=mkconf mkimage-iso-efi mkimage-raw-efi mkverification-raw-efi mkrootfs-ext4 mkrootfs-squash external-boot-image
+PKGS_DOCKER_LOAD=mkconf mkimage-iso-efi mkimage-raw-efi mkverification-raw-efi mkrootfs-ext4 mkrootfs-squash
 # these packages should exists for HOSTARCH as well as for ZARCH
 # alpine-base, alpine and cross-compilers are dependencies for others
 PKGS_HOSTARCH=alpine-base alpine cross-compilers $(PKGS_DOCKER_LOAD)
@@ -757,12 +757,9 @@ pkgs: RESCAN_DEPS=
 pkgs: build-tools $(PKGS)
        @echo Done building packages
 
-pkg/external-boot-image: eve-external-boot-image
-       docker tag $(shell $(LINUXKIT) pkg show-tag pkg/external-boot-image) lfedge/eve-external-boot-image:latest
-       docker save -o pkg/kube/external-boot-image.tar lfedge/eve-external-boot-image:latest
-       $(QUIET): $@: Succeeded
-pkg/kube: pkg/external-boot-image eve-kube
-       rm -f pkg/kube/external-boot-image.tar
+pkg/kube: eve-external-boot-image
+       $(MAKE) cache-export IMAGE=$(shell $(LINUXKIT) pkg show-tag pkg/external-boot-image) OUTFILE=pkg/kube/external-boot-image.tar && \
+       $(MAKE) eve-kube && rm -f pkg/kube/external-boot-image.tar
        $(QUIET): $@: Succeeded
 pkg/pillar: pkg/dnsmasq pkg/gpt-tools pkg/dom0-ztools eve-pillar
        $(QUIET): $@: Succeeded
```

File: pkg/external-boot-image/Dockerfile (hunk 2 above)
Comment (rene) [reply pkg/external-boot-image/Dockerfile:4]: yes, I agree, use KERNEL_TAG would be better....

File: pkg/external-boot-image/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+
+FROM scratch
+COPY --from=initrd-build --chmod=666 /usr/lib/xen/boot/runx-initrd /
+COPY --from=kernel-build --chmod=666 /kernel /
```
Comment (rene) [inline pkg/external-boot-image/Dockerfile:8]: @zedi-pramodh , I realized that you forgot to copy the kernel modules, you are copying only the kernel and initrd images, you need to copy the modules as well...

Comment (zedi-pramodh) [reply pkg/external-boot-image/Dockerfile:8]: Actually kubevirt only takes kernel and initrd.   Look at  https://kubevirt.io/user-guide/virtual_machines/boot_from_external_source/

File: Makefile (hunk 1 above)
Comment (zedi-pramodh) [reply Makefile:760]: Thanks, I will try this out and see how it goes.

File: pkg/external-boot-image/Dockerfile (hunk 3 above)
Comment (rene) [reply pkg/external-boot-image/Dockerfile:8]: ok, but in this case you will need to handle the modules anyways, otherwise the VM might not work properly... maybe create a new initrd and put the modules inside... or does kubevirt handles modules somehow outside this container?

Comment (zedi-pramodh) [reply pkg/external-boot-image/Dockerfile:8]: I have to guess so, because VMs are running fine and ztests passed. I also logged into VMs and verified things are fine.

**Review by eriknordmark — APPROVED** [review]
LGTM

File: pkg/external-boot-image/Dockerfile (hunk 2 above)
Comment (zedi-pramodh) [reply pkg/external-boot-image/Dockerfile:4]: @rene do you know how to pass env variables like KERNEL_TAG to linuxkit  pkg build ?  docker build seems to have an option but not linuxkit build.

File: Makefile (hunk 1 above)
Comment (zedi-pramodh) [reply Makefile:760]: This worked but I cannot pass in the env variables to linuxkit build.

File: pkg/external-boot-image/Dockerfile (hunk 3 above)
Comment (rene) [reply pkg/external-boot-image/Dockerfile:8]: It makes sense because you are using our initrd (from pkg/xen-tools), which mounts the module dir to the container, but at the other hand I did't understand where and how exactly this container is initialized.... but you said it's working fine, so have you checked the modules, they are there (under /lib/modules/....), right?

**Comment by rene** [comment]
> Could we avoid "Fix Yetus" commits? It's much cleaner to do `git commit --amend --no-edit` or a proper `git rebase -i` if necessary...

I agree, in this case the Yetus errors fixed were introduced by the changes in the PR, so all the commits can be safely squashed into the first one (Generate external boot image)...

File: pkg/external-boot-image/Dockerfile (hunk 3 above)
Comment (zedi-pramodh) [reply pkg/external-boot-image/Dockerfile:8]: This container is initialized in domainmgr code. We pass it to kubevirt in the spec. You will see that code in future commits. But the idea is described in the URL I posted earlier.

https://kubevirt.io/user-guide/virtual_machines/boot_from_external_source/

File: pkg/external-boot-image/Dockerfile (hunk 2 above)
Comment (zedi-pramodh) [reply pkg/external-boot-image/Dockerfile:4]: Ok I will go ahead and add a script under tools to process the KERNEL_TAG and as well as XENTOOLS_TAG for pkg external-boot-image

**Comment by zedi-pramodh** [comment]
Sure will collapse all commits to  one.

**Comment by zedi-pramodh** [comment]
@rene I believe I addressed all review comments and also collapsed everything to just one commit. Please review.
Verified pkg/kube gets compiled.

**Comment by zedi-pramodh** [comment]
Not sure how to address the Yetus issues reported in pkg/external-boot-image/Dockerfile. The ARG variable itself has the image tag.

File: pkg/external-boot-image/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# syntax=docker/dockerfile-upstream:1.5.0-rc2-labs
+
+ARG KERNEL=
+ARG XENTOOLS=
+
+#FROM lfedge/eve-xen-tools:e36677ad441c2d5cd88cdd9831959eaa4715ab82 as initrd-build
```
Comment (rene) [inline pkg/external-boot-image/Dockerfile:6]: You can remove these commented lines

File: Makefile [hunk 5]
```diff
@@ -757,6 +757,13 @@ pkgs: RESCAN_DEPS=
 pkgs: build-tools $(PKGS)
 	@echo Done building packages
 
+compose-external-boot-image:
```
Comment (rene) [inline Makefile:760]: You don't need to use a PHONY target here and generate the build.yml all the time. You can use build.yml as the target and build.yml.in as source, like the following:

```make
pkg/external-boot-image/build.yml: pkg/external-boot-image/build.yml.in
```

Also, compose-external-boot-image is not a dependency for any package, so if the build.yml file is not present at pkg/external-boot-image folder, it will fail if I hit manually `make pkg/external-boot-image` before run `make compose-external-boot-image`

File: pkg/external-boot-image/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+#FROM docker.io/lfedge/eve-kernel:eve-kernel-amd64-v6.1.38-generic-d2da815271a2-gcc as  kernel-build
+FROM $XENTOOLS as initrd-build
```
Comment (rene) [inline pkg/external-boot-image/Dockerfile:8]: @zedi-pramodh , we never used this approach before because we want to always fix the container version in order to keep the reproducibility of the build system. However, now that the kernel comes from outside our main tree, we knew this would happen sooner or later. Please, take a look on these links, where this approach was also discussed:

- https://github.com/moby/moby/issues/18119
- https://github.com/moby/moby/pull/31352

I think their approach is better and it can address the Yetus hadoling issue. The changes are minimal, you just need to use something like this (parsing the variables accordingly):

```Dockerfile
ARG KERNEL=latest
ARG XENTOOLS=latest

FROM lfedge/eve-xen-tools:$XENTOOLS as initrd-build
FROM lfedge/eve-kernel:$KERNEL as kernel-build
```

**Comment by rene** [comment]
> Could we avoid "Fix Yetus" commits? It's much cleaner to do `git commit --amend --no-edit` or a proper `git rebase -i` if necessary...

I agree, @zedi-pramodh you can squash 

> @rene I believe I addressed all review comments and also collapsed everything to just one commit. Please review. Verified pkg/kube gets compiled.

Thanks @zedi-pramodh for the update. I've added some comments and did some tests. Please, let me propose the following changes (that will address my comments and the Yetus issues). Feel free to use them (or not):

```diff
diff --git a/Makefile b/Makefile
index 62bff74d0..948b6ada7 100644
--- a/Makefile
+++ b/Makefile
@@ -757,9 +757,10 @@ pkgs: RESCAN_DEPS=
 pkgs: build-tools $(PKGS)
        @echo Done building packages
 
-compose-external-boot-image:
-       $(QUIET)tools/compose-external-boot-image-yml.sh pkg/external-boot-image/build.yml.in pkg/external-boot-image/build.yml ${KERNEL_TAG} $(shell $(LINUXKIT) pkg show-tag pkg/xen-tools) 
-pkg/kube: compose-external-boot-image
+pkg/external-boot-image/build.yml: pkg/external-boot-image/build.yml.in
+       $(QUIET)tools/compose-external-boot-image-yml.sh $< $@ $(shell echo ${KERNEL_TAG} | cut -d':' -f2) $(shell $(LINUXKIT) pkg show-tag pkg/xen-tools | cut -d':' -f2)
+pkg/external-boot-image: pkg/external-boot-image/build.yml
+pkg/kube: pkg/external-boot-image
        $(MAKE) eve-external-boot-image
        $(MAKE) cache-export IMAGE=$(shell $(LINUXKIT) pkg show-tag pkg/external-boot-image) OUTFILE=pkg/kube/external-boot-image.tar && \
        $(MAKE) eve-kube && rm -f pkg/kube/external-boot-image.tar && rm -f pkg/external-boot-image/build.yml
diff --git a/pkg/external-boot-image/Dockerfile b/pkg/external-boot-image/Dockerfile
index b358a2617..c429d2c4b 100644
--- a/pkg/external-boot-image/Dockerfile
+++ b/pkg/external-boot-image/Dockerfile
@@ -1,12 +1,10 @@
 # syntax=docker/dockerfile-upstream:1.5.0-rc2-labs
 
-ARG KERNEL=
-ARG XENTOOLS=
+ARG KERNEL=latest
+ARG XENTOOLS=latest
 
-#FROM lfedge/eve-xen-tools:e36677ad441c2d5cd88cdd9831959eaa4715ab82 as initrd-build
-#FROM docker.io/lfedge/eve-kernel:eve-kernel-amd64-v6.1.38-generic-d2da815271a2-gcc as  kernel-build
-FROM $XENTOOLS as initrd-build
-FROM $KERNEL as kernel-build
+FROM lfedge/eve-xen-tools:$XENTOOLS as initrd-build
+FROM lfedge/eve-kernel:$KERNEL as kernel-build
 
 FROM scratch
 COPY --from=initrd-build --chmod=666 /usr/lib/xen/boot/runx-initrd /
```

**Comment by zedi-pramodh** [comment]
Thanks @rene I will try these suggested changes which will fix Yetus too.

**Comment by zedi-pramodh** [comment]
@rene I pretty much took your diff and applied. I made just one additional change in Makefile

Added external-boot-image to be ignored on non kubevirt builds. Though it compiles and works, I do not see any reason to even compile that pkg on non-kubevirt builds.

PKGS_$(ZARCH)=$(shell find pkg -maxdepth 1 -type d | grep -Ev "eve|alpine|sources|kube|external-boot-image|verification$$")

**Comment by zedi-pramodh** [comment]
@eriknordmark I see PR builds passed for all processor types. I see only 1 check failed and 17 successful.

**Review by eriknordmark — APPROVED** [review]
Lgtm

**Review by rene — APPROVED** [review]
Thanks @zedi-pramodh , LGTM.

**Review by eriknordmark — APPROVED** [review]
Kick to run eden again (since a simple re-run from the workflow page results in the jobs being cancelled for some unknown reason)
